cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+{
+    Bundle bundle = {.name = "package_global"};
+    PromiseType promise_type = {.name = "package_global",
+                               .parent_bundle = &bundle};
+    Promise pp = {.promiser = "package_global",
+                  .parent_promise_type = &promise_type};
+    
+    CfLock package_promise_global_lock;
+    
+    package_promise_global_lock =
+            AcquireLock(ctx, GLOBAL_PACKAGE_PROMISE_LOCK_NAME, VUQNAME, CFSTARTTIME,
+                        (TransactionContext) {.ifelapsed = VIFELAPSED,
+                                              .expireafter = VEXPIREAFTER},
+                        &pp, false);
+                        
+    return (PackagePromiseGlobalLock) {.g_lock = package_promise_global_lock,
+                                       .lock_ctx = ctx};
+}
+
+void YieldGlobalPackagePromiseLock(PackagePromiseGlobalLock lock)
+{
+    Bundle bundle = {.name = "package_global"};
+    PromiseType promise_type = {.name = "package_global",
+                               .parent_bundle = &bundle};
+    Promise pp = {.promiser = "package_global",
+                  .parent_promise_type = &promise_type};
+    
+    YieldCurrentLockAndRemoveFromCache(lock.lock_ctx, lock.g_lock,
+                                       GLOBAL_PACKAGE_PROMISE_LOCK_NAME, &pp);
+}
+
+static void ParseAndLogErrorMessage(const Rlist *data)
+{
+    for (const Rlist *rp = data; rp != NULL; rp = rp->next)
+    {
+        char *line = RlistScalarValue(rp);
+                   
+        if (StringStartsWith(line, "Error="))
+        {
+            Log(LOG_LEVEL_VERBOSE, "Caught error: %s", line + strlen("Error="));
+        }
+        else if (StringStartsWith(line, "ErrorMessage="))
+        {
+            Log(LOG_LEVEL_VERBOSE, "Caught error message: %s", 
+                line + strlen("ErrorMessage="));
+        }
+        else
+        {
+            Log(LOG_LEVEL_VERBOSE, "Caught unsupported error info: %s", line);
+        }
+    }
+}
+
+static int IsReadWriteReady(const IOData *io, int timeout_sec)
+{
+    fd_set  rset;
+    FD_ZERO(&rset);
+    FD_SET(io->read_fd, &rset);
+
+    struct timeval tv = {
+        .tv_sec = timeout_sec,
+        .tv_usec = 0,
+    };
+
+    //TODO: For Windows we will need different method and select might not 
+    //      work with file descriptors.
+    int ret = select(io->read_fd + 1, &rset, NULL, NULL, &tv);
+
+    if(ret < 0)
+    {
+        Log(LOG_LEVEL_VERBOSE, "Failed checking for data. (select: %s)",
+            GetErrorStr());
+        return -1;
+    }
+
+    else if (FD_ISSET(io->read_fd, &rset))
+    {
+        return io->read_fd;
+    }
+
+    /* We have reached timeout */
+    if(ret == 0)
+    {
+        Log(LOG_LEVEL_DEBUG, "Timeout reading from package module.");
+        return 0;
+    }
+
+    Log(LOG_LEVEL_VERBOSE,
+        "Unknown outcome (ret > 0 but our only fds is not set). (select: %s)",
+        GetErrorStr());
```

Comment:
`GetErrorStr()` shall, at best, return something irrelevant, which you shouldn't attribute to `select`, as it succeeded so won't have set `errno`.